wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java

```diff
@@ -445,65 +439,117 @@ private MethodSpec messageEquals(MessageType type) {
-      if (afterFirstAssignment) {
-        result.addStatement("result = result * 37 + ($L != null ? $L.hashCode() : $L)",
-            name, name, nullHashValue(field));
+      Type.Name fieldType = field.type();
+      boolean isEnum = javaGenerator.isEnum(fieldType);
+      CodeBlock adapter = typeAdapter(fieldType);
+
+      if (field.isRepeated()) {
+        if (field.isPacked()) {
+          if (fieldType.isScalar()) {
+            result.addStatement("case $1L: builder.$2L = reader.packed(builder.$2L, $3L); break",
+                tag, name, adapter);
+          } else {
+            TypeName fieldName = javaGenerator.typeName(fieldType);
+            result.addStatement(
+                "case $1L: builder.$2L = packedEnumOrUnknown($1L, builder.$2L, reader, $3T.ADAPTER, builder); break",
+                tag, name, fieldName);
+          }
+        } else {
+          if (fieldType.isScalar()) {
+            result.addStatement(
+                "case $1L: builder.$2L = repeated(builder.$2L, reader.value($3L)); break", tag,
+                name, adapter);
+          } else {
+            TypeName fieldName = javaGenerator.typeName(fieldType);
+            if (isEnum) {
+              result.addStatement(
+                  "case $1L: builder.$2L = repeatedEnumOrUnknown($1L, builder.$2L, reader, $3T.ADAPTER, builder); break",
+                  tag, name, fieldName);
+            } else {
+              result.addStatement(
+                  "case $1L: builder.$2L = repeatedMessage(builder.$2L, reader, $3T.ADAPTER); break",
+                  tag, name, fieldName);
+            }
+          }
+        }
+      } else if (fieldType.isScalar()) {
+        result.addStatement("case $L: builder.$L = reader.value($L); break", tag, name, adapter);
       } else {
-        result.addStatement("result = $L != null ? $L.hashCode() : $L",
-            name, name, nullHashValue(field));
-        afterFirstAssignment = true;
+        TypeName fieldName = javaGenerator.typeName(fieldType);
+        if (isEnum) {
+          result.addStatement(
+              "case $1L: builder.$2L = enumOrUnknown($1L, reader, $3T.ADAPTER, builder); break",
+              tag, name, fieldName);
+        } else {
+          result.addStatement("case $1L: builder.$2L = message(reader, $3T.ADAPTER); break", tag,
+              name, fieldName);
+        }
       }
     }
-    result.addStatement("hashCode = result");
-    result.endControlFlow();
-    result.addStatement("return result");
+
+    if (!type.extensions().isEmpty()) {
+      result.addStatement("default: builder.readExtensionOrUnknown(tag, reader); break");
+    } else {
+      result.addStatement("default: builder.readUnknown(tag, reader); break");
+    }
+
+    result.endControlFlow(); // Switch
+    result.endControlFlow(); // While
+    result.addStatement("return builder.build()");
+
     return result.build();
   }
 
+  private CodeBlock typeAdapter(Type.Name fieldName) {
```

Comment:
Maybe we should rename `Type.Name` to `Type.Named` so the noun is unambiguously "type" rather than "name" (which is probably what I kept thinking during refactors).